addons/print/models/print_printer.py

```diff
@@ -23,80 +23,82 @@ def _find_lpr_exec():
-    parent_right = fields.Integer(string="Right parent", index=True)
+    is_group = fields.Boolean(string="Is Printer Group", default=False)
+    group_id = fields.Many2one(
+        "print.printer",
+        string="Printer Group",
+        ondelete="cascade",
+        domain=[("is_group", "=", True)],
+    )
+    child_ids = fields.One2many("print.printer", "group_id", string="Grouped Printers")
+    parent_path = fields.Char(index=True)
 
     _sql_constraints = [
-        ('barcode_uniq', 'unique (barcode)', "The Barcode must be unique"),
-        ('single_default',
-         'exclude (is_default with =) where (is_default and group_id is null)',
-         "There must be only one System Default Printer"),
-        ('single_group_default',
-         'exclude (group_id with =) where (is_default)',
-         "There must be only one System Default Printer per group"),
+        ("barcode_uniq", "unique (barcode)", "The Barcode must be unique"),
+        (
+            "single_default",
+            "exclude (is_default with =) where (is_default and group_id is null)",
+            "There must be only one System Default Printer",
+        ),
+        (
+            "single_group_default",
+            "exclude (group_id with =) where (is_default)",
+            "There must be only one System Default Printer per group",
+        ),
     ]
 
-    @api.multi
-    @api.depends('name', 'group_id.full_name')
+    @api.depends("name", "group_id.full_name")
     def _compute_full_name(self):
         """Calculate full name (including group name(s))"""
         for printer in self:
             if printer.group_id.full_name:
-                printer.full_name = '%s / %s' % (
-                    printer.group_id.full_name, printer.name
-                )
+                printer.full_name = "%s / %s" % (printer.group_id.full_name, printer.name)
             else:
                 printer.full_name = printer.name
 
-    @api.multi
     def _compute_is_user_default(self):
         """Calculate user default flag"""
         for printer in self:
             printer.is_user_default = printer in self.env.user.printer_ids
 
-    @api.multi
-    @api.constrains('is_group', 'group_id', 'child_ids')
+    @api.constrains("is_group", "group_id", "child_ids")
     def _check_groups(self):
         """Constrain group existence"""
         for printer in self:
             if printer.group_id and not printer.group_id.is_group:
-                raise ValidationError(_("%s is not a printer group") %
-                                      printer.group_id.name)
+                raise ValidationError(_("%s is not a printer group") % printer.group_id.name)
             if printer.child_ids and not printer.is_group:
-                raise ValidationError(_("%s is not a printer group") %
-                                      printer.name)
+                raise ValidationError(_("%s is not a printer group") % printer.name)
+
+    @api.model
+    def set_default_printer(self):
```

Comment:
I don't get it. Isn't this function resetting the default printer if no printer is active? But Why is it only present in the xml file in data? What's the point of this function? I'm a little confused.

Reply:
When the module is installed or updated it makes sure that the system has a default printer set. If it doesn't then it will be set to the one that is included with the module.

But having said that, I don't see the `default_printer` XML record...

Reply:
@PeterC10 https://github.com/unipartdigital/odoo-print/blob/story/x1178/addons/print/data/print_printer_data.xml